setup-container.sh

```diff
@@ -1,176 +1,377 @@
+        exit_failure "failed to generate SSH key pair: $(basename "${PRIVKEY_FILE}")/$(basename "${PUBKEY_FILE}")"
     fi
 
-    echo "Verifying UID and GID in container matches host"
-    CONTAINER_USER_ID=`docker exec $CONTAINER_NAME sh -c "id $USER | grep -o \"uid=[0-9]*\" | cut -d \"=\" -f 2"`
-    CONTAINER_GROUP_ID=`docker exec $CONTAINER_NAME sh -c "id $USER | grep -o \"gid=[0-9]*\" | cut -d \"=\" -f 2"`
+    log_info "read SSH public key: ${PUBKEY_FILE}"
+    SSH_PUBKEY="$(cat "${PUBKEY_FILE}")" || \
+    exit_failure "failed to read public SSH key: ${PUBKEY_FILE}"
 
-    if [[ "$HOST_USER_ID" != "$CONTAINER_USER_ID" ]]; then
-        echo "User ID mismatch between host and container"
-        exit 1
+    if [[ -f "${AUTHKEY_FILE}" ]]; then
+        grep -q "${SSH_PUBKEY}" "${AUTHKEY_FILE}" || echo "${SSH_PUBKEY}" >> "${AUTHKEY_FILE}"
+    else
+        echo "${SSH_PUBKEY}" > "${AUTHKEY_FILE}"
+        chmod 0644 "${AUTHKEY_FILE}"
     fi
 
-    if [[ "$HOST_GROUP_ID" != "$CONTAINER_GROUP_ID" ]]; then
-        echo "Group ID mismatch between host and container"
-        exit 1
+    TMP_DIR="$(mktemp -d)"
+    log_info "setup a temporary dir: ${TMP_DIR}"
+
+    log_info "copy SSH key pair: $(basename "${PRIVKEY_FILE}")/$(basename "${PUBKEY_FILE}")"
+    eval "cp -fv \"${PRIVKEY_FILE}\" \"${TMP_DIR}/id_rsa\" ${SILENT_HOOK}"
+    eval "cp -fv \"${PUBKEY_FILE}\" \"${TMP_DIR}/id_rsa.pub\" ${SILENT_HOOK}"
+
+    log_info "prepare a Dockerfile template: ${TMP_DIR}/Dockerfile.j2"
+    cat <<'EOF' > "${TMP_DIR}/Dockerfile.j2"
+FROM {{ IMAGE_ID }}
+
+USER root
+
+# Group configuration
+RUN if getent group {{ GROUP_NAME }}; \
+then groupmod -o -g {{ GROUP_ID }} {{ GROUP_NAME }}; \
+else groupadd -o -g {{ GROUP_ID }} {{ GROUP_NAME }}; \
+fi
+
+# User configuration
+RUN if getent passwd {{ USER_NAME }}; \
+then usermod -o -g {{ GROUP_ID }} -u {{ USER_ID }} -m -d /home/{{ USER_NAME }} {{ USER_NAME }}; \
+else useradd -o -g {{ GROUP_ID }} -u {{ USER_ID }} -m -d /home/{{ USER_NAME }} -s /bin/bash {{ USER_NAME }}; \
+fi
+
+# Docker configuration
+RUN if getent group {{ DGROUP_NAME }}; \
+then groupmod -o -g {{ DGROUP_ID }} {{ DGROUP_NAME }}; \
+else groupadd -o -g {{ DGROUP_ID }} {{ DGROUP_NAME }}; \
+fi
+
+# Environment configuration
+RUN if [ '{{ USER_NAME }}' != 'AzDevOps' ]; then \
+/bin/bash -c 'cp -a -f /var/AzDevOps/* /home/{{ USER_NAME }}/'; \
+/bin/bash -c 'cp -a -f /var/AzDevOps/{.profile,.local,.ssh} /home/{{ USER_NAME }}/'; \
+fi
+
+# Permissions configuration
+RUN usermod -a -G sudo {{ USER_NAME }}
+RUN usermod -a -G {{ DGROUP_NAME }} {{ USER_NAME }}
+RUN echo '{{ USER_NAME }} ALL=(ALL) NOPASSWD:ALL' > /etc/sudoers.d/{{ USER_NAME }}
+RUN chmod 0440 /etc/sudoers.d/{{ USER_NAME }}
+RUN chown -R '{{ USER_ID }}:{{ GROUP_ID }}' /home/{{ USER_NAME }}
+
+# SSH/PASS configuration
+RUN sed -i -E 's/^#?PermitRootLogin.*$/PermitRootLogin yes/g' /etc/ssh/sshd_config
+RUN echo 'root:{{ ROOT_PASS }}' | chpasswd
```

Comment:
What is the use case for adding root login?

Reply:
@theasianpianist this is done for management purposes only: on some older docker versions there is no automatic terminal adjusting according to the window size, so using the exec is a bit painful. This is a reason of having SSH for root.